### PR TITLE
Remaining unit tests for `submitRedemptionProof`

### PR DESCRIPTION
**Pull request by lukasz-zimnoch**
Refs: #118 
Depends on: #153 

This change covers all test scenarios for `submitRedemptionProof` function.

### CODE REVIEW

File: solidity/test/bridge/Bridge.test.ts [hunk 1]
```diff
@@ -2233,229 +2252,1840 @@ describe("Bridge", () => {
...
+
+                          // Alter the last redemption requests in the test
+                          // data and set such an amount that will cause
+                          // the Bitcoin redemption transaction to be deemed
+                          // as invalid due to a wrong amount. The corresponding
+                          // transaction output has the value of 191169 so to
+                          // make this test scenario happen, the request amount
+                          // must be way different (lesser or greater) than the
+                          // output value.
+                          data.redemptionRequests[4].amount = 100000
```
Comment (pdyraga) [inline solidity/test/bridge/Bridge.test.ts:3715]: Just like in the cost of a single output, I think we could improve tests by adding two separate cases:
- requested amount - treasury fee - max Bitcoin transaction fee - 1
- requested amount + 1

We would then have all corner cases covered.

File: solidity/test/bridge/Bridge.test.ts [hunk 2]
```diff
@@ -2233,229 +2252,1840 @@ describe("Bridge", () => {
...
+                        it("should revert", async () => {
+                          await expect(outcome).to.be.revertedWith(
+                            "Output value is not within the acceptable range of the pending request"
+                          )
+                        })
+                      }
+                    )
+
+                    context(
+                      "when output vector contains a reported timed out requested redemption with wrong amount",
```
Comment (pdyraga) [inline solidity/test/bridge/Bridge.test.ts:3733]:
```suggestion
                      "when output vector contains a reported timed out requested redemption with wrong amount redeemed",
```

File: solidity/test/bridge/Bridge.test.ts [hunk 3]
```diff
@@ -2233,229 +2252,1840 @@ describe("Bridge", () => {
...
+                        it("should revert", async () => {
+                          await expect(outcome).to.be.revertedWith(
+                            "Output value is not within the acceptable range of the timed out request"
+                          )
+                        })
+                      }
+                    )
+
+                    context(
+                      "when output vector contains a non-zero P2SH change output",
```
Comment (pdyraga) [inline solidity/test/bridge/Bridge.test.ts:3784]: We expect it to revert because P2SH-type main UTXO is not allowed and only P2PKH is valid, correct?

https://developer.bitcoin.org/devguide/transactions.html says:

> As of Bitcoin Core 0.9, the standard pubkey script types are:
> - Pay To Public Key Hash (P2PKH)
> - Pay To Script Hash (P2SH)
> - Multisig
> - Pubkey
> - Null Data

Should we have similar cases for the last three? What about witness?

**Review by pdyraga — COMMENTED** [review]
Solid work @lukasz-zimnoch! 🚀 

Left a couple of non-blocking questions, I am fine with merging as-is but will wait for your replies in case they are game-changers.

File: solidity/test/bridge/Bridge.test.ts [hunk 4]
```diff
@@ -2233,229 +2252,1840 @@ describe("Bridge", () => {
...
+
+                          // Alter the single redemption request in the test
+                          // data and set such an amount that will cause
+                          // the Bitcoin redemption transaction to be deemed
+                          // as invalid due to a wrong amount. The transaction
+                          // output has the value of 1176924 so to make this
+                          // test scenario happen, the request amount must be
+                          // way different (lesser or greater) than the output
+                          // value.
+                          data.redemptionRequests[0].amount = 300000
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/test/bridge/Bridge.test.ts:2513]: Yeah but I think we have quite good documentation in the contract explaining how the correct redemption range is defined. I had to use some simplification in this test scenario because it's not so straightforward to test both edge cases at once since the requested amount is also validated against the wallet balance which is determined using an actual testnet BTC UTXO. That means we need to assemble two separate and quite tricky BTC transactions and test data for it. This is why I just tested this case by reusing `SinglePendingRequestedRedemption` data set and just checked the amount condition in a general way.

File: solidity/test/bridge/Bridge.test.ts [hunk 5]
```diff
@@ -2233,229 +2252,1840 @@ describe("Bridge", () => {
...
+
+                        after(async () => {
+                          await restoreSnapshot()
+                        })
+
+                        // We have this case because P2SH script has a 20-byte
+                        // payload which may match the 20-byte wallet public
+                        // key hash though it should be always rejected as
+                        // non-requested output. There is no need to check for
+                        // P2WSH since the payload is always 32-byte there.
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/test/bridge/Bridge.test.ts:2656]: Yep. We only need to bother about 20-bytes hashes. This is because the wallet public key hash has always 20 bytes and we must make sure no redemption request uses it as a redeemer script to not confuse an output that will try to handle that request with a proper change output also referencing the wallet public key hash.

[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/test/bridge/Bridge.test.ts:2656]: Can you add your response to my question to the end of the comment for this test? This clarifies a lot.

File: solidity/test/data/redemption.ts [hunk 6]
```diff
@@ -73,18 +73,568 @@ export interface RedemptionBalanceChange {
...
+  chainDifficulty: 1,
+}
+
+/**
+ * `SingleNonRequestedRedemption` test data represents a redemption with
+ *  the following properties:
+ * - 1 input pointing to the wallet main UTXO
+ * - 0 redemption requests
+ * - 1 P2SH output with value 1176924 satoshi and index 0, pointing to a
+ *   non-requested output script.
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/test/data/redemption.ts:153]: Yeah but this is just a comment that describes the data set. What happens with that data set and all consequences should be defined at a specific unit test level.

File: solidity/test/bridge/Bridge.test.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/test/bridge/Bridge.test.ts:2656]: Sure!

File: solidity/test/data/redemption.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/test/data/redemption.ts:153]: Right. What I mean is that `SinglePendingRequestedRedemption` specifies wallet's pubkey hash and explain there is no change output. Here we know there is one output and that the redemption was not requested but we do not know (just looking at the docs) if that output is pointing back to the wallet or not. As far as I understand, it's not something we can set up in the test.

File: solidity/test/bridge/Bridge.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/test/bridge/Bridge.test.ts:2513]: I see... Can you add a simplified version of this explanation to the comment? Makes sense to me and I think it's a good-to-know for future-us.

File: solidity/test/bridge/Bridge.test.ts [hunk 7]
```diff
@@ -2233,229 +2252,1840 @@ describe("Bridge", () => {
...
+                            redeemerBalance.afterProof.sub(
+                              redeemerBalance.beforeProof
+                            )
+                          ).to.be.equal(0, "Balance of redeemer has changed")
+                        })
+                      }
+                    )
+
+                    context(
+                      "when the single output is a pending requested redemption but amount is wrong",
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/test/bridge/Bridge.test.ts]: See https://github.com/keep-network/tbtc-v2/pull/155/commits/d83e9046d941b5fd2d31d9d688ccb0c4f04ea0c3

File: solidity/test/bridge/Bridge.test.ts [hunk 8]
```diff
@@ -2233,229 +2252,1840 @@ describe("Bridge", () => {
...
+
+                          outcome = runRedemptionScenario(data)
+                        })
+
+                        after(async () => {
+                          await restoreSnapshot()
+                        })
+
+                        // Should be deemed as valid change though rejected
+                        // because this change is a single output.
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/test/bridge/Bridge.test.ts]: See https://github.com/keep-network/tbtc-v2/pull/155/commits/d83e9046d941b5fd2d31d9d688ccb0c4f04ea0c3

File: solidity/test/bridge/Bridge.test.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/test/bridge/Bridge.test.ts:2658]: See https://github.com/keep-network/tbtc-v2/pull/155/commits/d83e9046d941b5fd2d31d9d688ccb0c4f04ea0c3

File: solidity/test/bridge/Bridge.test.ts [hunk 9]
```diff
@@ -2233,229 +2252,1840 @@ describe("Bridge", () => {
...
+                              0,
+                              `Balance of redeemer with index ${i} has changed`
+                            )
+                          }
+                        })
+                      }
+                    )
+
+                    context(
+                      "when output vector contains a pending requested redemption with wrong amount",
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/test/bridge/Bridge.test.ts]: See https://github.com/keep-network/tbtc-v2/pull/155/commits/d83e9046d941b5fd2d31d9d688ccb0c4f04ea0c3

File: solidity/test/bridge/Bridge.test.ts (hunk 2 above)
Comment (lukasz-zimnoch) [reply solidity/test/bridge/Bridge.test.ts]: See https://github.com/keep-network/tbtc-v2/pull/155/commits/d83e9046d941b5fd2d31d9d688ccb0c4f04ea0c3

File: solidity/test/bridge/Bridge.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/test/bridge/Bridge.test.ts]: See https://github.com/keep-network/tbtc-v2/pull/155/commits/66841439336e820b3e582905bb08be09458e02e5

File: solidity/test/bridge/Bridge.test.ts (hunk 1 above)
Comment (lukasz-zimnoch) [reply solidity/test/bridge/Bridge.test.ts:3739]: And same response as in https://github.com/keep-network/tbtc-v2/pull/155#discussion_r821661775

File: solidity/test/data/redemption.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/test/data/redemption.ts]: Yeah right, got your point now. The fact this output does not go back to the wallet is implied by `pointing to a non-requested output script.` because a change output is always an expected one. However, I agree this can be not obvious at a first sight. Added an explicit explanation: https://github.com/keep-network/tbtc-v2/pull/155/commits/94be16ef1b22933b1897710b838145131a7d7bb3

File: solidity/test/bridge/Bridge.test.ts [hunk 10]
```diff
@@ -2233,229 +2252,1840 @@ describe("Bridge", () => {
...
+                              0,
+                              `Balance of redeemer with index ${i} has changed`
+                            )
+                          }
+                        })
+                      }
+                    )
+
+                    context(
+                      "when output vector consists only of reported timed out requested redemptions",
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/test/bridge/Bridge.test.ts:3030]: Nope, we have this case covered only for one output in "when the single output is a non-reported timed out requested redemption" test case. I think it is redundant to check it for multiple outputs. Technically this is the same as just multiple pending requests and we already confirmed it works for one output within the aforementioned case.

**Review by pdyraga — APPROVED** [review]
